File: 2024/14/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n");

class Robot {
  constructor(line) {
    const numbers = line.match(/-?\d+/g).map(Number);
    this.pos = [numbers[0], numbers[1]];
    this.vel = [numbers[2], numbers[3]];
  }
  simulate(time, spaceWidth, spaceHeight) {
    for (let i = 0; i < time; i++) {
      this.step(spaceWidth, spaceHeight);
    }
    return this.pos;
  }
  step(spaceWidth, spaceHeight) {
    this.pos[0] += this.vel[0];
    this.pos[1] += this.vel[1];
    if (this.pos[0] < 0) this.pos[0] += spaceWidth;
    else this.pos[0] %= spaceWidth;
    if (this.pos[1] < 0) this.pos[1] += spaceHeight;
    else this.pos[1] %= spaceHeight;
    return this.pos;
  }
}

const spaceWidth = 101;
const spaceHeight = 103;
const halfWidth = (spaceWidth - 1) / 2;
const halfHeight = (spaceHeight - 1) / 2;
const robots = input.map((line) => new Robot(line));
const quads = [0, 0, 0, 0];
robots.forEach((robot) => {
  const [x, y] = robot.simulate(100, spaceWidth, spaceHeight);
  if (x < halfWidth) {
    if (y < halfHeight) quads[0]++;
    else if (y > halfHeight) quads[1]++;
  } else if (x > halfWidth) {
    if (y < halfHeight) quads[2]++;
    else if (y > halfHeight) quads[3]++;
  }
});
console.log(quads.reduce((a, b) => a * b));

let i = 100;
while (true) {
  i++;
  const coords = new Set();
  robots.forEach((robot) => {
    const [x, y] = robot.step(spaceWidth, spaceHeight);
    coords.add(`${x},${y}`);
  });
  if (coords.size === 500) {
    console.log(i);
    console.log("\n");
    const canvas = Array(spaceHeight)
      .fill()
      .map(() => Array(spaceWidth).fill(" "));
    coords.forEach((coord) => {
      const [x, y] = coord.split(",").map(Number);
      canvas[y][x] = "#";
    });
    console.log(canvas.map((row) => row.join("")).join("\n"));
    break;
  }
}

File: 2024/14/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	fmt.Println(part1(parseInput(strings.Split(string(data), "\n"))))
}

const (
	WIDTH  = 101
	HEIGHT = 103
	RMAX   = HEIGHT - 1
	CMAX   = WIDTH - 1
	RHALF  = RMAX / 2
	CHALF  = CMAX / 2
)

type Robot struct {
	p [2]int
	v [2]int
}

var robotRegex = regexp.MustCompile(`^p=(-?\d+),(-?\d+) v=(-?\d+),(-?\d+)$`)

func NewRobot(s string) *Robot {
	match := robotRegex.FindStringSubmatch(s)
	if match == nil {
		panic(fmt.Sprintf("Error matching robot regex with: %s", s))
	}
	var nums [4]int
	for i := range 4 {
		n, err := strconv.Atoi(match[i+1])
		if err != nil {
			panic(err)
		}
		nums[i] = n
	}
	return &Robot{[2]int{nums[0], nums[1]}, [2]int{nums[2], nums[3]}}
}

func (r *Robot) Quadrant() int {
	if r.p[0] < CHALF {
		if r.p[1] < RHALF {
			return 1
		} else if r.p[1] > RHALF {
			return 2
		}
	} else if r.p[0] > CHALF {
		if r.p[1] < RHALF {
			return 3
		} else if r.p[1] > RHALF {
			return 4
		}
	}
	return 0
}

func (r *Robot) Simulate(k int) {
	r.p[0] = (r.p[0] + k*r.v[0]) % WIDTH
	if r.p[0] < 0 {
		r.p[0] += WIDTH
	}
	r.p[1] = (r.p[1] + k*r.v[1]) % HEIGHT
	if r.p[1] < 0 {
		r.p[1] += HEIGHT
	}
}

func parseInput(puzzleInput []string) []*Robot {
	robots := make([]*Robot, len(puzzleInput))
	for i, s := range puzzleInput {
		robots[i] = NewRobot(s)
	}
	return robots
}

func part1(robots []*Robot) int {
	var quadrants [5]int
	for _, robot := range robots {
		robot.Simulate(100)
		quadrants[robot.Quadrant()]++
	}
	res := 1
	for i := range 4 {
		res *= quadrants[i+1]
	}
	return res
}
